docs(en-US): sign in with Logto on Spring Boot (#231)
logto-io/docs: docs/docs/recipes/integrate-logto/spring-boot.mdx

Follow the options below:

- Gradle Project
- Language: Java
- Spring Boot: 2.7.2

Generate and open the project.

## Add dependencies

Add the dependencies to your Gradle project build file `build.gradle`:

```groovy
// path/to/project/build.gradle
dependencies {
    implementation("com.fasterxml.jackson.datatype:jackson-datatype-jsr310")
    implementation("org.springframework.boot:spring-boot-starter-web")
    implementation("org.springframework.boot:spring-boot-starter-security")
    implementation("org.springframework.boot:spring-boot-starter-oauth2-client")
    implementation("org.springframework.boot:spring-boot-starter-thymeleaf")
    implementation("org.thymeleaf.extras:thymeleaf-extras-springsecurity5")
    implementation("nz.net.ultraq.thymeleaf:thymeleaf-layout-dialect")
}
```

:::note

Since Spring Boot and Spring Security have built-in support for both OAuth 2 and OIDC,
you DO NOT need to add additional libraries from Logto to integrate.

See [Spring Security and OpenID Connect](https://www.baeldung.com/spring-security-openid-connect)
and [Spring Security Architecture](https://spring.io/guides/topicals/spring-security-architecture)

> for details if you're interested.

Reviewer: ```suggestion
for more details.
```

:::

## Configure Spring Security

Spring Security makes it easy to configure your application for authentication with OIDC providers such as Logto.

> Uses an `application.yml` file (instead of the default `application.properties`) to configure the OAuth2 client and provider.

Reviewer: ```suggestion
Use an `application.yml` file (instead of the default `application.properties`) to configure the OAuth2 client and provider.
```

```yaml
# path/to/project/src/main/resources/application.yaml
spring:
  security:
    oauth2:
      client:
        registration:
          logto:
            client-id: <your-app-id>

>             client-secret: <your-app-secret> # Copied from application details in Admin Console

Reviewer: ```suggestion
            client-secret: <your-app-secret> # Copy from application details in Admin Console
```

            scope:
              - openid

>               - profile

Reviewer 2: also `offline_access`

>               - profile

Author: ```suggestion
              - profile
              - offline_access
```

Reviewer 2: Since we add `offline_access` in scopes, ensure the sign-in URI contains `prompt=consent` and the fetched token response contains a refresh token.

        provider:
          logto:

>             issuer-uri: <your-issuer-uri> # e.g. `http://localhost:3001/oidc`. You can replace it with your logto's production issuer URI.

Reviewer: ```suggestion
            issuer-uri: <your-issuer-uri> # e.g. `http://localhost:3001/oidc`. Replace it with your logto issuer URI.
```

import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
import org.springframework.security.oauth2.client.web.DefaultOAuth2AuthorizationRequestResolver;
import org.springframework.security.oauth2.client.web.OAuth2AuthorizationRequestCustomizers;
import org.springframework.security.oauth2.client.web.OAuth2AuthorizationRequestResolver;
import org.springframework.security.web.SecurityFilterChain;

@EnableWebSecurity
public class SecurityConfiguration {

	@Bean
	public OAuth2AuthorizationRequestResolver oauth2AuthorizationRequestResolverWithPkce(
			ClientRegistrationRepository clientRegistrationRepository
	) {
		DefaultOAuth2AuthorizationRequestResolver authorizationRequestResolver =
				new DefaultOAuth2AuthorizationRequestResolver(clientRegistrationRepository, "/oauth2/authorization");
		authorizationRequestResolver.setAuthorizationRequestCustomizer(OAuth2AuthorizationRequestCustomizers.withPkce());
		return  authorizationRequestResolver;
	}

	@Bean
	public SecurityFilterChain securityFilterChain(
			HttpSecurity http,
			OAuth2AuthorizationRequestResolver oauth2AuthorizationRequestResolverWithPkce
	) throws Exception {

> 		http.authorizeRequests()
> 		    // Anyone can access the home page.
> 		    // (At least one mapping is required. Change it as you need.)
>         .mvcMatchers("/").permitAll()
> 				.and().oauth2Login()
> 				.authorizationEndpoint()
> 				.authorizationRequestResolver(oauth2AuthorizationRequestResolverWithPkce);

Author: ```suggestion
		http.authorizeRequests()
		        // Anyone can access the home page.
		        // (At least one mapping is required. Change it as you need.)
                        .mvcMatchers("/").permitAll()
                        .and().oauth2Login()
                        .authorizationEndpoint()
				.authorizationRequestResolver(oauth2AuthorizationRequestResolverWithPkce);
```

		return http.build();
	}
}
```

</details>

> ### Implement sign-in in route

Reviewer: ```suggestion
### Implement sign-in route
```

Spring Security will use the `spring.security.oauth2.client` configuration defined above to handle user sign-in automatically.

Let's try this sign-in link on your web application page.

Add a home page template file:

```html
<!-- path/to/project/src/main/resources/templates/index.html -->
<!DOCTYPE html>
<html
  lang="en"
  xmlns:th="http://www.thymeleaf.org"
  xmlns:sec="http://www.thymeleaf.org/thymeleaf-extras-springsecurity5"
>
  <body>
    <div sec:authorize="!isAuthenticated()">
      <a th:href="@{/oauth2/authorization/logto}">Sign in</a>
    </div>
    <div sec:authorize="isAuthenticated()">
      <p>Signed in</p>
    </div>
  </body>
</html>
```

Add a controller to render the view:

```java
package io.logto.springboot.sample.controller;

import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;

@Controller
public class HomeController {

    @GetMapping("/")
    public String home() {
        return "index";
    }
}
```

Run your Spring Boot web application, visit the home page [http://localhost:3000/](http://localhost:3000/), click the _Sign in_ link on it, and it will redirect to the Logto sign-in page.

> After the user signs in successfully, it will redirect back to the home page showing "Signed in" that means the user is authenticated.

Reviewer: ```suggestion
Provide your credentials, and once authenticated, it will redirect you back to the home page with a message that shows "Signed in".
```
